tests/unit/test_synapse_workload.py

```diff
@@ -72,6 +72,98 @@ def test_allow_public_rooms_over_federation_error(monkeypatch: pytest.MonkeyPatc
+    """
+    arrange: set mock container with file.
+    act: update ip_range_whitelist config and call enable_ip_range_whitelist.
+    assert: new configuration file is pushed and ip_range_whitelist is enabled.
+    """
+    root = harness.get_filesystem_root(synapse.SYNAPSE_CONTAINER_NAME)
+    config_path = root / synapse.SYNAPSE_CONFIG_PATH[1:]
+    config_path.parent.mkdir(parents=True, exist_ok=True)
+    config_path.write_text(
+        """
+listeners:
+    - type: http
+      port: 8080
+      bind_addresses:
+        - "::"
+"""
+    )
+
+    container: ops.Container = harness.model.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+    harness.update_config({"ip_range_whitelist": ip_range_whitelist})
+    harness.begin()
+    synapse.enable_ip_range_whitelist(container, harness.charm._charm_state)
+
+    with open(config_path, encoding="utf-8") as config_file:
+        content = yaml.safe_load(config_file)
+        expected_config_content = {
+            "listeners": [
+                {"type": "http", "port": 8080, "bind_addresses": ["::"]},
+            ],
+            "ip_range_whitelist": synapse.workload._create_ip_range_whitelist(
+                harness.charm._charm_state
+            ),
+        }
+        assert content == expected_config_content
+
+
+def test_enable_ip_range_whitelist_error(harness: Harness, monkeypatch: pytest.MonkeyPatch):
+    """
+    arrange: set mock container with file.
+    act: update ip_range_whitelist config and call enable_smtp.
+    assert: raise WorkloadError in case of error.
+    """
+    error_message = "Error pulling file"
+    path_error = ops.pebble.PathError(kind="fake", message=error_message)
+    pull_mock = MagicMock(side_effect=path_error)
+    container_mock = MagicMock()
+    monkeypatch.setattr(container_mock, "pull", pull_mock)
+
+    with pytest.raises(synapse.WorkloadError, match=error_message):
+        expected_first_domain = "foo1"
+        expected_second_domain = "foo2"
+        harness.update_config(
+            {"ip_range_whitelist": f"{expected_first_domain},{expected_second_domain}"}
+        )
+        harness.begin()
+        synapse.enable_ip_range_whitelist(container_mock, harness.charm._charm_state)
+
+
+def test_enable_ip_range_whitelist_no_action(harness: Harness, monkeypatch: pytest.MonkeyPatch):
+    """
+    arrange: set mock container with file.
+    act: leave ip_range_whitelist config empty and call enable_ip_range_whitelist.
+    assert: configuration file is not changed.
+    """
+    config_content = """
+    listeners:
+        - type: http
+          port: 8080
+          bind_addresses:
+            - "::"
+    """
+    text_io_mock = io.StringIO(config_content)
+    pull_mock = Mock(return_value=text_io_mock)
+    push_mock = MagicMock()
+    container_mock = MagicMock()
+    monkeypatch.setattr(container_mock, "pull", pull_mock)
+    monkeypatch.setattr(container_mock, "push", push_mock)
```

Comment:
Similar to above, could we just construct one MagicMock and rely on that attributes are automatically generated? That way we save a few lines here